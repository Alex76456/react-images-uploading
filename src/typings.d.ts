export interface ImageType {
  dataURL: string;
  file?: File;
  key?: string;
  onUpdate?: () => void;
  onRemove?: () => void;
}

export type ImageListType = Array<ImageType>;

export interface ImageUploadingPropsType {
  children?: (props: ExportInterface) => React.ReactNode;
  defaultValue?: ImageListType;
  onChange?: (value: ImageListType) => void;
  multiple?: boolean;
  maxNumber?: number;
  acceptType?: Array<string>;
  maxFileSize?: number;
}

export interface ExportInterface {
  imageList: ImageListType;
  onImageUpload: () => void;
  onImageRemoveAll: () => void;
  errors: Record<string, any>;
}

export type ErrorsType = {
  maxFileSize: boolean;
  maxNumber: boolean;
  acceptType: boolean;
};
