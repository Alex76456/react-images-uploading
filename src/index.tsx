import * as React from "react";
import { getBase64 } from "./utils";

const { useRef, useState, useEffect } = React;
export interface ImageType {
  dataURL: string;
  file?: File;
  key?: string;
  onUpdate?: () => void;
  onRemove?: () => void;
}

export type ImageListType = Array<ImageType>;

export interface ImageUploadingPropsType {
  children?: (props: ExportInterface) => React.ReactNode;
  defaultValue?: ImageListType;
  onChange?: (value: ImageListType) => void;
  multiple?: boolean;
  maxNumber?: number;
  acceptType?: Array<string>;
  maxFileSize?: number;
}

export interface ExportInterface {
  imageList: ImageListType;
  onImageUpload: () => void;
  onImageRemoveAll: () => void;
  errors: Record<string, any>;
}

type ErrorsType = {
  maxFileSize: boolean;
  maxNumber: boolean;
  acceptType: boolean;
};

const defaultErrors: ErrorsType = {
  maxFileSize: false,
  maxNumber: false,
  acceptType: false,
};

const ImageUploading: React.FC<ImageUploadingPropsType> = ({
  multiple,
  onChange,
  maxNumber,
  children,
  defaultValue,
  acceptType,
  maxFileSize,
}) => {
  const inputRef = useRef<HTMLInputElement>(null);
  const [imageList, setImageList] = useState(() => {
    let initImageList: Array<ImageType> = [];
    if (defaultValue) {
      initImageList = defaultValue.map((item: ImageType) => ({
        ...item,
        key: item.dataURL,
        onUpdate: (): void => onImageUpdate(item.dataURL),
        onRemove: (): void => onImageRemove(item.dataURL),
      }));
    }
    return initImageList;
  });

  const [keyUpdate, setKeyUpdate] = useState<string>("");
  const [errors, setErrors] = useState<ErrorsType>({ ...defaultErrors });

  const imageListRef = useRef(imageList);
  useEffect(() => {
    imageListRef.current = imageList;
  }, [imageList]);

  const onStandardizeDataChange = (list: ImageListType): void => {
    if (onChange) {
      const sData: ImageListType = list.map((item) => ({
        file: item.file,
        dataURL: item.dataURL,
      }));
      onChange(sData);
    }
  };

  const onImageUpload = (): void => {
    inputRef.current && inputRef.current.click();
  };

  const onImageRemoveAll = (): void => {
    setImageList([]);
    onStandardizeDataChange([]);
  };

  const onImageRemove = (key: string): void => {
    const updatedList: ImageListType = imageListRef.current.filter(
      (item: ImageType) => item.key !== key
    );
    setImageList(updatedList);
    onStandardizeDataChange(updatedList);
  };

  useEffect(() => {
    if (keyUpdate) {
      onImageUpload();
    }
  }, [keyUpdate]);

  const onImageUpdate = (key: string): void => {
    setKeyUpdate(key);
  };

  const getListFile = (files: FileList): Promise<ImageListType> => {
    const promiseFiles: Array<Promise<string>> = [];

    for (let i = 0; i < files.length; i++) {
      promiseFiles.push(getBase64(files[i]));
    }

    return Promise.all(promiseFiles).then((fileListBase64: Array<string>) => {
      const fileList: ImageListType = fileListBase64.map((base64, index) => {
        const key = `${new Date().getTime().toString()}-${files[index].name}`;
        return {
          dataURL: base64,
          file: files[index],
          key,
          onUpdate: (): void => onImageUpdate(key),
          onRemove: (): void => onImageRemove(key),
        };
      });
      return fileList;
    });
  };

  const validate = (fileList: ImageListType): boolean => {
    setErrors({ ...defaultErrors });
    if (maxNumber && fileList.length + imageList.length > maxNumber) {
      setErrors({ ...errors, maxNumber: true });
      return false;
    }
    if (maxFileSize || (acceptType && acceptType.length > 0)) {
      for (let i = 0; i < fileList.length; i++) {
        const file = fileList[i].file;
        if (file) {
          if (maxFileSize) {
            const size = Math.round(file.size / 1024 / 1024);
            if (size > maxFileSize) {
              setErrors({ ...errors, maxFileSize: true });
              return false;
            }
          } else if (acceptType) {
            const type: string = file.name.split(".").pop() || "";
            if (acceptType.indexOf(type) < 0) {
              setErrors({ ...errors, acceptType: true });
              return false;
            }
          }
        }
      }
    }
    return true;
  };

  const onInputChange = async (
    e: React.ChangeEvent<HTMLInputElement>
  ): Promise<void> => {
    const { files } = e.target;

    if (files) {
      const fileList = await getListFile(files);
      if (fileList.length > 0) {
        if (validate(fileList)) {
          let updatedFileList: ImageListType;
          if (keyUpdate) {
            updatedFileList = imageList.map((item: ImageType) => {
              if (item.key === keyUpdate) return { ...fileList[0] };
              return item;
            });
            setKeyUpdate("");
          } else {
            if (multiple) {
              updatedFileList = [...imageList, ...fileList];
              if (maxNumber && updatedFileList.length > maxNumber) {
                updatedFileList = imageList;
              }
            } else {
              updatedFileList = [fileList[0]];
            }
          }
          setImageList(updatedFileList);
          onStandardizeDataChange(updatedFileList);
        } else {
        }
      } else {
        keyUpdate && setKeyUpdate("");
      }
    }
  };

  const acceptString =
    acceptType && acceptType.length > 0
      ? acceptType.map((item) => `.${item}`).join(", ")
      : "image/*";

  return (
    <>
      <input
        type="file"
        accept={acceptString}
        ref={inputRef}
        multiple={multiple && !keyUpdate}
        onChange={onInputChange}
        style={{ display: "none" }}
      />
      {children &&
        children({
          imageList,
          onImageUpload,
          onImageRemoveAll,
          errors,
        })}
    </>
  );
};

ImageUploading.defaultProps = {
  maxNumber: 1000,
  multiple: false,
  acceptType: [],
};

export default ImageUploading;
