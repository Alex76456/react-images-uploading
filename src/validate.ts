import { DEFAULT_NULL_INDEX } from './constants';
import { ResolutionType } from './typings';

export const isResolutionValid = (
  dataURL: string,
  resolutionType: ResolutionType = 'absolute',
  resolutionWidth: number = 0,
  resolutionHeight: number = 1
): Promise<boolean> => {
  const image = new Image();

  return new Promise((resolve) => {
    image.addEventListener('load', () => {
      if (image.width && image.height) {
        switch (resolutionType) {
          case 'absolute': {
            if (
              image.width === resolutionWidth &&
              image.height === resolutionHeight
            )
              return resolve(true);
            return resolve(false);
          }
          case 'ratio': {
            const ratio = resolutionWidth / resolutionHeight;
            if (image.width / image.height === ratio) return resolve(true);
            return resolve(false);
          }
          case 'less': {
            if (
              image.width <= resolutionWidth &&
              image.height <= resolutionHeight
            )
              return resolve(true);
            return resolve(false);
          }
          case 'more': {
            if (
              image.width >= resolutionWidth &&
              image.height >= resolutionHeight
            )
              return resolve(true);
            return resolve(false);
          }
        }
      }
      resolve(true);
    });
    image.src = dataURL;
  });
};

export const isImageValid = (fileType: string) => {
  if (fileType.includes('image')) {
    return true;
  }
  return false;
};

export const maxFileSizeValid = (fileSize, maxFileSize) => {
  return maxFileSize ? fileSize <= maxFileSize : true;
};

export const acceptTypeValid = (acceptType, fileName) => {
  if (acceptType && acceptType.length > 0) {
    const type: string = fileName.name.split('.').pop() || '';
    if (acceptType.indexOf(type) < 0) {
      return false;
    }
  }
  return true;
};

export const maxNumberValid = (totalNumber, maxNumber, keyUpdate) => {
  if (keyUpdate === DEFAULT_NULL_INDEX && maxNumber && totalNumber > maxNumber)
    return false;
  return true;
};
