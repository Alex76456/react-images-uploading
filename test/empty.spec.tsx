import * as React from 'react';
import { mount } from 'enzyme';
import { stub } from 'sinon';
import toJson from 'enzyme-to-json';
import { ImageListType } from '../src';
import { App } from './fixtures/app';

describe('testing empty case', () => {
  it('should render component', () => {
    const wrapper = mount(
      <App value={[] as ImageListType} onChange={stub()} />
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
