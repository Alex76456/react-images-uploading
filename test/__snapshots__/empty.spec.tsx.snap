// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing empty case should render component 1`] = `
<Component
  onChange={[Function]}
  value={Array []}
>
  <div
    id="app"
  >
    <ReactImageUploading
      onChange={[Function]}
      value={Array []}
    >
      <input
        accept="image/*"
        multiple={false}
        onChange={[Function]}
        style={
          Object {
            "display": "none",
          }
        }
        type="file"
      />
      <div
        className="upload__image-wrapper"
      >
        <button
          id="btn-upload"
          onClick={[Function]}
          onDragEnter={[Function]}
          onDragLeave={[Function]}
          onDragOver={[Function]}
          onDrop={[Function]}
        >
          Click or Drop here
        </button>
        <button
          id="btn-remove"
          onClick={[Function]}
        >
          Remove all images
        </button>
        <div
          id="list-images"
        />
      </div>
    </ReactImageUploading>
  </div>
</Component>
`;
